akka-sample-sharding-scala/src/main/scala/sample/sharding/WeatherEdges.scala

```diff
@@ -0,0 +1,145 @@
+package sample.sharding
+
+import scala.concurrent.duration._
+import scala.util.{ Failure, Random, Success }
+
+import akka.actor.typed.{ ActorRef, ActorSystem, Behavior }
+import akka.actor.typed.scaladsl.{ Behaviors, TimerScheduler }
+import com.typesafe.config.ConfigFactory
+
+/** Simulate devices and stations.
+ * In the wild, each station would run its own simple system,
+ * taking samples and posting via HTTPS.
+ */
+object WeatherEdges {
+
+  private val WeatherStations = 50
+
+  // temperature, pressure, precipitation, wind speed...
+  private val DevicesPerStation = 1
+
+  val random = new Random()
+
+  def main(args: Array[String]): Unit = {
+    val config = ConfigFactory.load("weather-station.conf")
+
+    // simple sample, using one node on local machine
+    val rootBehavior = Behaviors.setup[Nothing] { context =>
+      (0 to WeatherStations).foreach { sid =>
+        context.spawn(Station(sid, DevicesPerStation, 8081), s"station-$sid")
+      }
+      Behaviors.empty
+    }
+
+    ActorSystem[Nothing](rootBehavior, "SimulatedDevices", config)
+  }
+}
+
+object Station {
+
+  def apply(stationId: Int, devices: Int, port: Int): Behavior[WeatherHttpApi.Data] =
+    Behaviors.setup { context =>
+      context.log.info("Started station with {} devices.", devices)
+      val http = context.spawn(WeatherHttpApi(stationId, port), s"http-post-$stationId")
+
+      (0 to devices).foreach { deviceId =>
+        context.spawn(WeatherDevice(stationId, deviceId, http, 1.second, 10), s"device-$deviceId")
+      }
+
+      Behaviors.empty
+    }
+}
+
+/**
+ * Any given weather station has a unique ID, and can sample temperature, wind speed, precipitation, etc.
+ * from multiple devices.
+ *
+ * For a simple example, this does one type of reading.
+ */
+object WeatherHttpApi {
+
+  final case class Data(stationId: Int, deviceId: Int, data: List[Double])
+
+  def apply(stationId: Int, port: Int): Behavior[Data] =
+    Behaviors.setup[Data] { context =>
+      import akka.http.scaladsl.Http
+      import akka.http.scaladsl.model.{ ContentTypes, HttpEntity, HttpMethods, HttpRequest }
+
+      implicit val sys = context.system
+      import context.executionContext
+      import akka.actor.typed.scaladsl.adapter._
+      val http = Http(context.system.toClassic)
+
+      // Run and completely consume a single akka http request
+      def runRequest(req: HttpRequest) =
+        http.singleRequest(req).flatMap { _.entity.dataBytes.runReduce(_ ++ _) }
+
+      Behaviors.receiveMessage[Data] {
+        case Data(sid, did, data) =>
+          // TODO marshalling vs scary hand-coded!
+          val values = data.mkString(",")
+          val json = s"""{"stationId":$sid,"deviceId":$did,"data":[$values]}"""
```

Comment:
@raboof this is my wall of (marshalling) shame.

Reply:
If you add a format for `Data` in `JsonFormats` that in the implicit scope I think this should be `Marshal(data).to[MessageEntity]`, see docs here: https://doc.akka.io/docs/akka-http/current/common/marshalling.html#using-marshallers

Reply:
Done